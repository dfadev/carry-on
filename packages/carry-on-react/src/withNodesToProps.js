/* eslint-disable react/jsx-props-no-spreading */
import React from "react";
import hoistNonReactStatic from "hoist-non-react-statics";

const nodesToProps = children => {
  const childs = Array.isArray(children) ? children : [children];

  const newProps = {};

  for (let i = 0, len = childs.length; i < len; i += 1) {
    const child = childs[i];
    /* eslint-disable-next-line no-continue */
    if (child === undefined || child === null) continue;

    // fetch node map config
    const {
      prop,
      val = "children",
      transform,
      default: def
    } = child.type || {};

    // retrieve value
    let v = child.props && child.props[val];

    // default value
    if (v === undefined) v = def;

    // transform value
    v = transform ? transform(v, child.props) : v;

    // present as array when multiple nodes
    const curProp = newProps[prop];
    if (curProp !== undefined) {
      if (Array.isArray(curProp)) {
        curProp.push(v);
      } else {
        newProps[prop] = [curProp, v];
      }
    } else newProps[prop] = v;
  }

  return newProps;
};

const withNodesToProps = WrappedComponent => {
  const WithNodesToProps = ({ children, ...props }) =>
    !Array.isArray(children) && typeof children === "function" ? (
      <WrappedComponent {...props}>{children}</WrappedComponent>
    ) : (
      <WrappedComponent {...props} {...nodesToProps(children)} />
    );

  WithNodesToProps.displayName = "withNodesToProps";
  WithNodesToProps.WrappedComponent = WrappedComponent;

  return hoistNonReactStatic(WithNodesToProps, WrappedComponent);
};

export default withNodesToProps;
