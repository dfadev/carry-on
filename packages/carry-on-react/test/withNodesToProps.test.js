/**
 * @jest-environment jsdom
 */
import React from "react";
import { render } from "@testing-library/react";
import withNodesToProps from "../src/withNodesToProps";

test("withNodesToProps", () => {
  const Inner = () => null;
  const PropVal = () => null;
  const Default = () => null;

  const nodeMap = {
    Inner: "inner",
    PropVal: { prop: "val", val: "val", default: 0 },
    Default: { prop: "default", default: true, transform: s => s }
  };

  const Element = withNodesToProps(nodeMap, ({ inner, val, default: def }) => (
    <div>{Array.isArray(inner) ? inner.join(", ") : inner} {val} {def ? "true" : "false"}</div>
  ));

  const { asFragment } = render(
    <Element>
      <Inner>{"stuff"}</Inner>
      <PropVal val="123" />
      <Default />
    </Element>
  );
  expect(asFragment()).toMatchSnapshot();

  const { asFragment: asFragment2 } = render(
    <Element>
      <Inner>{"stuff"}</Inner>
      <Inner>{"stuff"}</Inner>
      <Inner>{"stuff"}</Inner>
      <PropVal val="123" />
      <PropVal val="123" />
      <PropVal val="123" />
      <Default />
    </Element>
  );
  expect(asFragment2()).toMatchSnapshot();
});
