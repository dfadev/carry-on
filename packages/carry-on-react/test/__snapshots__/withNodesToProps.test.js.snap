// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withNodesToProps 1`] = `
<DocumentFragment>
  <div>
    stuff 123 true
  </div>
</DocumentFragment>
`;

exports[`withNodesToProps 2`] = `
<DocumentFragment>
  <div>
    stuff, stuff, stuff 123123123 true
  </div>
</DocumentFragment>
`;
